### ISSUE TITLE
Errors on GRPC PHP - pecl install grpc

### ISSUE THREAD

**Issue by bshaffer** [grpc/grpc#12109]
As of today, I am getting this error [on travis](https://travis-ci.org/GoogleCloudPlatform/php-docs-samples/jobs/261942015#L1140) for the command `pecl install grpc`: 

```
$ travis_retry pecl install grpc
# ...
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c: At top level:
cc1: error: unrecognized command line option "-Wno-parentheses-equality" [-Werror]
cc1: all warnings being treated as errors
make: *** [src/php/ext/grpc/channel.lo] Error 1
ERROR: `make' failed
```

This error seems to be related to https://github.com/grpc/grpc/issues/7571, which was closed almost a year ago.

I only receive this error on PHP 5.6, but I am receiving a different error when running grpc tests on PHP 7.0. The build [times out after 10 minutes](https://travis-ci.org/GoogleCloudPlatform/php-docs-samples/jobs/261994466#L5824), and the logs reveal the following error:

```
E0807 19:31:09.399814488    6759 sync_posix.c:60]            assertion failed: pthread_mutex_lock(mu) == 0
6759 Aborted                 (core dumped) vendor/bin/phpunit

[secure]_time:end:16586363:start=1502133989762474252,finish=1502134269616946002,duration=279854471750
```

Here is the full command output for `pecl install grpc`

```
_retry pecl install grpc
downloading grpc-1.4.3.tgz ...
Starting to download grpc-1.4.3.tgz (2,301,057 bytes)
..................done: 2,301,057 bytes
1048 source files, building
running: phpize
Configuring for:
PHP Api Version:         20131106
Zend Module Api No:      20131226
Zend Extension Api No:   220131226
building in /tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3
running: /tmp/pear/temp/grpc/configure --with-php-config=/home/[secure]/.phpenv/versions/5.6.24/bin/php-config
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... x86_64-unknown-linux-gnu
checking for PHP prefix... /home/[secure]/.phpenv/versions/5.6.24
checking for PHP includes... -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib
checking for PHP extension directory... /home/[secure]/.phpenv/versions/5.6.24/lib/php/extensions/no-debug-zts-20131226
checking for PHP installed headers prefix... /home/[secure]/.phpenv/versions/5.6.24/include/php
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... re2c
checking for re2c version... 0.13.5 (ok)
checking for gawk... gawk
checking whether to enable grpc support... yes, shared
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking for ld used by cc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for /usr/bin/ld option to reload object files... -r
checking for BSD-compatible nm... /usr/bin/nm -B
checking whether ln -s works... yes
checking how to recognize dependent libraries... pass_all
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking dlfcn.h usability... yes
checking dlfcn.h presence... yes
checking for dlfcn.h... yes
checking how to run the C++ preprocessor... g++ -E
checking the maximum length of command line arguments... 1572864
checking command to parse /usr/bin/nm -B output from cc object... ok
checking for objdir... .libs
checking for ar... ar
checking for ranlib... ranlib
checking for strip... strip
checking if cc supports -fno-rtti -fno-exceptions... no
checking for cc option to produce PIC... -fPIC
checking if cc PIC flag -fPIC works... yes
checking if cc static flag -static works... yes
checking if cc supports -c -o file.o... yes
checking whether the cc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no

creating libtool
appending configuration tag "CXX" to libtool
checking for ld used by g++... /usr/bin/ld -m elf_x86_64
checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC
checking if g++ PIC flag -fPIC works... yes
checking if g++ static flag -static works... yes
checking if g++ supports -c -o file.o... yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... GNU/Linux ld.so
(cached) (cached) checking how to hardcode library paths into programs... immediate
configure: creating ./config.status
config.status: creating config.h
running: make
/bin/bash /tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/libtool --mode=compile cc -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include  -DHAVE_CONFIG_H  -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11   -c /tmp/pear/temp/grpc/src/php/ext/grpc/byte_buffer.c -o src/php/ext/grpc/byte_buffer
.lo
mkdir src/php/ext/grpc/.libs
 cc -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include -DHAVE_CONFIG_H -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11 -c /tmp/pear/temp/grpc/src/php/ext/grpc/byte_buffer.c  -fPIC -DPIC -o src/php/ext/grpc/.libs/byte_buffer.o
/bin/bash /tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/libtool --mode=compile cc -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include  -DHAVE_CONFIG_H  -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11   -c /tmp/pear/temp/grpc/src/php/ext/grpc/call.c -o src/php/ext/grpc/call.lo
 cc -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include -DHAVE_CONFIG_H -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11 -c /tmp/pear/temp/grpc/src/php/ext/grpc/call.c  -fPIC -DPIC -o src/php/ext/grpc/.libs/call.o
/bin/bash /tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/libtool --mode=compile cc -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include  -DHAVE_CONFIG_H  -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11   -c /tmp/pear/temp/grpc/src/php/ext/grpc/call_credentials.c -o src/php/ext/grpc/call_c
redentials.lo
 cc -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include -DHAVE_CONFIG_H -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11 -c /tmp/pear/temp/grpc/src/php/ext/grpc/call_credentials.c  -fPIC -DPIC -o src/php/ext/grpc/.libs/call_credentials.o
/bin/bash /tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/libtool --mode=compile cc -fvisibility=hidden     -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN     -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include  -DHAVE_CONFIG_H  -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11   -c /tmp/pear/temp/grpc/src/php/ext/grpc/channel.c -o src/php/ext/grpc/channel.lo
 cc -fvisibility=hidden -DOPENSSL_NO_ASM -D_GNU_SOURCE -DWIN32_LEAN_AND_MEAN -D_HAS_EXCEPTIONS=0 -DNOMINMAX -DGRPC_ARES=0 -I. -I/tmp/pear/temp/grpc -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/include -I/tmp/pear/temp/pear-build-[secure]Vzvqp1/grpc-1.4.3/main -I/tmp/pear/temp/grpc -I/home/[secure]/.phpenv/versions/5.6.24/include/php -I/home/[secure]/.phpenv/versions/5.6.24/include/php/main -I/home/[secure]/.phpenv/versions/5.6.24/include/php/TSRM -I/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext -I/home/[secure]/.phpenv/versions/5.6.24/include/php/ext/date/lib -I/tmp/pear/temp/grpc/include -I/tmp/pear/temp/grpc/src/php/ext/grpc -I/tmp/pear/temp/grpc/third_party/boringssl/include -DHAVE_CONFIG_H -Wall -Werror -Wno-parentheses-equality -Wno-unused-value -std=c11 -c /tmp/pear/temp/grpc/src/php/ext/grpc/channel.c  -fPIC -DPIC -o src/php/ext/grpc/.libs/channel.o
In file included from /home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/zend.h:286:0,
                 from /home/[secure]/.phpenv/versions/5.6.24/include/php/main/php.h:35,
                 from /tmp/pear/temp/grpc/src/php/ext/grpc/channel.h:41,
                 from /tmp/pear/temp/grpc/src/php/ext/grpc/channel.c:34:
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c: In function â€˜create_and_add_channel_to_persistent_listâ€™:
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/../TSRM/TSRM.h:164:57: error: â€˜tsrm_lsâ€™ undeclared (first use in this function)
 #define TSRMG(id, type, element) (((type) (*((void ***) tsrm_ls))[TSRM_UNSHUFFLE_RSRC_ID(id)])->element)
                                                         ^
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/zend_hash.h:110:28: note: in definition of macro â€˜zend_hash_updateâ€™
   _zend_hash_add_or_update(ht, arKey, nKeyLength, pData, nDataSize, pDest, HASH_UPDATE ZEND_FILE_LINE_CC)
                            ^
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c:188:3: note: in expansion of macro â€˜PHP_GRPC_PERSISTENT_LIST_UPDATEâ€™
   PHP_GRPC_PERSISTENT_LIST_UPDATE(&EG(persistent_list), key, key_len,
   ^
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/zend_globals_macros.h:45:16: note: in expansion of macro â€˜TSRMGâ€™
 # define EG(v) TSRMG(executor_globals_id, zend_executor_globals *, v)
                ^
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c:188:36: note: in expansion of macro â€˜EGâ€™
   PHP_GRPC_PERSISTENT_LIST_UPDATE(&EG(persistent_list), key, key_len,
                                    ^
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/../TSRM/TSRM.h:164:57: note: each undeclared identifier is reported only once for each function it appears in
 #define TSRMG(id, type, element) (((type) (*((void ***) tsrm_ls))[TSRM_UNSHUFFLE_RSRC_ID(id)])->element)
                                                         ^
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/zend_hash.h:110:28: note: in definition of macro â€˜zend_hash_updateâ€™
   _zend_hash_add_or_update(ht, arKey, nKeyLength, pData, nDataSize, pDest, HASH_UPDATE ZEND_FILE_LINE_CC)
                            ^
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c:188:3: note: in expansion of macro â€˜PHP_GRPC_PERSISTENT_LIST_UPDATEâ€™
   PHP_GRPC_PERSISTENT_LIST_UPDATE(&EG(persistent_list), key, key_len,
   ^
/home/[secure]/.phpenv/versions/5.6.24/include/php/Zend/zend_globals_macros.h:45:16: note: in expansion of macro â€˜TSRMGâ€™
 # define EG(v) TSRMG(executor_globals_id, zend_executor_globals *, v)
                ^
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c:188:36: note: in expansion of macro â€˜EGâ€™
   PHP_GRPC_PERSISTENT_LIST_UPDATE(&EG(persistent_list), key, key_len,
                                    ^
/tmp/pear/temp/grpc/src/php/ext/grpc/channel.c: At top level:
cc1: error: unrecognized command line option "-Wno-parentheses-equality" [-Werror]
cc1: all warnings being treated as errors
make: *** [src/php/ext/grpc/channel.lo] Error 1
ERROR: `make' failed
```

**Comment by bshaffer** [comment]
I was able to run `pecl install grpc-1.4.3` on Mac OS X (PHP 5.6), but I received the same error once I ran the same tests

```
E0807 16:42:26.605299000 140736785834944 sync_posix.c:60]              assertion failed: pthread_mutex_lock(mu) == 0
/usr/local/bin/phpunit: line 2:  9547 Abort trap: 6           /usr/bin/env php -d allow_url_fopen=On -d detect_unicode=Off /usr/local/Cellar/phpunit/5.3.2/libexec/phpunit-5.3.2.phar "$@"
```

**Comment by bshaffer** [comment]
I received the same error on Mac OSX for PHP 7.0 and PHP 7.1

**Comment by stanley-cheung** [comment]
There are 2 issues here (with 1.4.3):
 - The Travis compilation error
 - The MacOS mutex segfault when running unit tests

Both will be fixed with #12108 soon

**Comment by anoopmv** [comment]
HI,

While running grpc hello world program I am facing this issue, please help
me in this:

Traceback (most recent call last):
  File "greeter_client.py", line 48, in <module>
    run()
  File "greeter_client.py", line 43, in run
    response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
  File "/usr/local/lib/python2.7/dist-packages/grpc/_channel.py", line 507,
in __call__
    return _end_unary_response_blocking(state, call, False, deadline)
  File "/usr/local/lib/python2.7/dist-packages/grpc/_channel.py", line 455,
in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with
(StatusCode.UNAVAILABLE, Connect Failed)>



Regards,
Anoop

**Comment by stanley-cheung** [comment]
@anoopmv What you mentioned was a Python issue. Please log a separate issue for that.

**Comment by anoopmv** [comment]
I raised this issue.

**Comment by michaelbausor** [comment]
I am seeing a similar error intermittently with the recently released grpc 1.4.4 from pecl, running PHP 5.6.25 on Ubuntu:

```
E0808 14:14:52.418892428    7743 sync_posix.c:60]            assertion failed: pthread_mutex_lock(mu) == 0
Aborted (core dumped)
```

**Comment by stanley-cheung** [comment]
Closing these thread as the original installation errors are gone. Now we are onto some specific tests when using the APIs.
